Feat: 카테고리별 도서 조회(정렬 & 페이징) 기능 구현

junga970 opened the pull request:
- #2

### 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
### **📌 AS-IS**
**1. 응답, 예외 처리**
- 기본적으로 ApiResponse으로 응답합니다.
- 예외 처리는 ResponseEntity<ErrorResponse>로 응답합니다.

<br>

**2. S3 버킷 생성 & 책 이미지 업로드 및 url 확인**
- (Book)image 컬럼에 url 저장
- spring boot와 연동된 상태는 아니고 직접 업로드한 상태입니다.

<br>

**2. 카테고리별 도서 조회 기능 추가**
- 쿼리 스트링을 사용하며 sub-category, order, page, size 값을 받습니다. (default값이 있어 필수는 아닙니다.)
- 서브카테고리가 존재하지 않을 경우 예외를 발생시킵니다.
- 최신순, 높은 가격순, 낮은 가격순, 판매순으로 정렬 가능합니다.
- 정렬 기준은 type/Order에 enum으로 정의했으며, 속성으로 Sort 객체를 정의합니다.
- 잘못된 order값인 경우 예외를 발생시킵니다.
- 조건에 맞는 책리스트를 반환합니다.

<br>

- 판매순으로 정렬하기 위해서 주문 내역 테이블을 미리 만들었습니다.
<br>


**3. 테스트 코드 추가 (통과된 테스트는 체크 표시)**

Service/BookServiceTest

- [x] 도서 조회 성공
- [x] 도서 조회 실패 - 존재하지 않는 서브카테고리
- [x] 도서 조회 실패 - 잘못된 정렬 값

<br>

### **📌 TO-BE**
1. 카테고리 생성 기능 추가

2. 서브카테고리 생성 기능 추가

3. 도서 등록 기능 추가
- spring boot 와 S3 연동


### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [x] 테스트 코드
- [x] API 테스트

junga970 committed:
Feat: 카테고리별 도서 조회(정렬 & 페이징) 기능 구현

- #2

mingming-mentor commented on src/main/java/com/example/bookstore/controller/BookController.java:
```diff
@@ -0,0 +1,38 @@
+
+	private final BookService bookService;
+
+	@GetMapping("/book")
```
명사같은 경우 s를 붙여서 복수형태로 사용하는 것이 Restful API의 naming convention중 하나입니다. 가능하면 따르시는 것을 추천드려요! 참고하실만한 블로그 첨부해드립니다 : https://prohannah.tistory.com/156

mingming-mentor commented on src/main/java/com/example/bookstore/controller/BookController.java:
```diff
@@ -0,0 +1,38 @@
+	@ResponseStatus(HttpStatus.OK)
+	public ApiResponse<List<BookInfo>> getBooksByCategory(
+		@RequestParam(value = "sub-category", defaultValue = "1") Long subCategoryId,
+		@RequestParam(value = "order", defaultValue = "bestseller") String orderValue,
```
Pageable이라는 클래스를 spring에서 제공하고 있는데 paging관련 처리는 이를 사용하셔도 좋을것 같아요

mingming-mentor commented on src/main/java/com/example/bookstore/repository/BookRepository.java:
```diff
@@ -0,0 +1,20 @@
+		+ "left join (SELECT book_id  , SUM(quantity) sales_quantity FROM order_book ob \n"
+		+ "GROUP BY book_id) b on b.id = b.book_id \n"
+		+ "where sc.id = :id", nativeQuery = true)
+	List<Book> findBySubCategoryId(@Param(value = "id") Long subCategoryId, Pageable pageable);
```
Pageable을 사용하는데 Page<Book>으로 받아올수는 없나용? (native라 안될지도 모른다는 생각이들긴 하는데 한번 여쭤봅니다..!)

junga970 replied:
```java
@Query(
		value = "SELECT * FROM book b\n"
			+ "LEFT JOIN (SELECT book_id, SUM(quantity) sales_quantity FROM order_book GROUP BY book_id) b\n"
			+ "ON b.id = b.book_id WHERE b.sub_category_id = :id",
		countQuery = "SELECT COUNT(*) FROM book WHERE sub_category_id = :id",
		nativeQuery = true
	)
```
countQuery 부분을 추가하니 Page로 받아와집니다!
혹시 물어보신 이유가 총 페이지 개수, 총 요소 개수 등등 때문일까요?

mingming-mentor commented on src/main/java/com/example/bookstore/service/BookService.java:
```diff
@@ -0,0 +1,68 @@
+
+	public static Order getOrder(String orderValue) {
+
+		for (Order order : Order.values()) {
```
Arrays.stream(Order.values()).filter(it -> it.getValue().equals(orderValue)).findFirst(); 로 좀더 간편히 사용할수 있구요. 요런 method의 경우 보통 현재 BookService와는 큰관계가 없다고 판단 Order class안에 구현해주고 있습니다..!

mingming-mentor commented on src/main/java/com/example/bookstore/entity/Book.java:
```diff
@@ -0,0 +1,43 @@
+
+	@Type(type = "json")
+	@Column(columnDefinition = "json")
+	private Map<String, List<String>> author;
```
작가가 어떤식으로 들어가게 되나요?

junga970 replied:
{"author":["작가1","작가2"]} 이렇게 들어가 있습니다.
작가가 여러명일 때를 고려해서 json 컬럼을 사용했지만
author컬럼에 author이라는 키값이 들어가는게 중복같아서...ㅎㅎ
String으로 바꾸고 값을 "작가1, 작가2" 이런식으로 넣을까 생각중입니다!
멘토님 생각은 어떠신지 궁금합니다 ㅎㅎ!

mingming-mentor replied:
답변이 늦었네요ㅠ 죄송합니다. 
아하! 이해했습니다. 저도 말씀하신대로 작가1, 작가2로 들어가면 좋을 것 같구요. 
json type이더라도 list로 바로 넣는것도 가능한걸로 알고 있어요. 즉 author column안에 
`{"author":["작가1","작가2"]}` 형태 말고도 `["작가1","작가2"]` 형태도 지원하는 걸로 알고 있습니다. https://dev.mysql.com/doc/refman/8.0/en/json.html

물론 string 형태도 괜찮을 것 같긴합니다. 다만 elastic search를 사용하지 않고 검색을 구현하실 경우, 작가 column을 string으로 하신다면 검색성능이 조금 떨어질 것 같아요. https://gngsn.tistory.com/163 요 블로그를 참고해보시면 좋을 것 같습니다.

junga970 replied:
좋은 정보 감사합니다! 작가 부분은 리스트로 변경했습니다 ㅎㅎ

mingming-mentor commented on src/main/java/com/example/bookstore/entity/Book.java:
```diff
@@ -0,0 +1,43 @@
+	private SubCategory subCategory;
+
+	private String title;
+	private String image;
```
imageUrl이라는 이름이 좀더 알맞을 것 같습니다

mingming-mentor reviewed (approved):
수고 많으셨습니다! 코드가 깔끔하고 가독성이 높아 리뷰하기가 좋았습니다 👍

junga970 committed:
Refactor: Book 관련 파일들 수정

1. Book, BookInfo: image -> imageUrl로 변경
- url임을 명시

2. BookController:
   1) size 파라미터 제거
   2) List<BookInfo> -> Page<BookInfo>로 변경,
   3) /book -> /books 매핑할 경로 변경
- size 고정, 총 페이지 개수, 총 요소 개수 등 정보를 넘기기 위해 Page 객체로 변경

3. BookRepository: 쿼리문 변경, Page 객체로 반환
- 불필요한 조인문 제거

4. BookService:
   1) getOrder 함수 제거 -> stream 사용,
   2) Sort 객체를 얻기 위한 if문 제거 -> Order.getSort() 사용
   3) Page 객체 반환으로 변경
- 가독성 고려

5. ErrorCode: 에러 코드 추가

6. SecurityConfig: /book -> /books로 경로 수정

7. BookServiceTest: 변경된 코드에 맞게 테스트 코드 수정

- #2

junga970 committed:
Refactor: Book, BookInfo 수정

Book, BookInfo: author을 Map에서 List로 변경
- 키가 필요 없음

- #2
